docs/design/SecretsEncryption.md

```diff
@@ -0,0 +1,192 @@
+# Encryption Master Keys
+
+## OVERVIEW
+Previously, all sensitive data (cloud pool credentials, chunk cipher keys, account sync credentials, etc.) was stored plain and un-encrypted in the DB. This poses a security threat, in which accessing the PVC of mongo would allow access to all the secrets and through that the data on the system and access to the cloud services.
+
+By encrypting these sensitive secrets, we can make sure that even if the PVC is compromised, the data and cloud access details won’t be.
+
+The suggestion is to use encryption keys, which will encrypt the secrets at rest (DB), for in-memory usage these secrets will be used in their decrypted format.
+
+## GOALS
+ * Encrypt/decrypt all secret keys in the DB.
+ * Maintain master keys per level (aka scope or tenant):
+    * root-master-key - A global key stored in KMS / HSM / Kubernetes secret (mounted through env), and used to encrypt the system level keys. 
+    * system-master-key - stored in the DB, encrypted by root-master-key, and used to encrypt the account/bucket level keys
+    * account-master-key - stored in the DB, encrypted by system-master-key, and used to encrypt all the cloud credentials and account S3 credentials.
+    * bucket-master-key - stored in the DB, encrypted by system-master-key, and used to encrypt all the chunks cipher keys of the bucket.
+ * Key rotation policies:
+    * On demand - API request to rotate all keys or a specific scope (root/system/account/bucket).
+    * Master keys expiry policy - set a system wide policy to expire and rotate keys automatically.
+    * Chunk keys expiry policy - set a bucket/system policy to expire the chunk cipher keys which will re-generate cipher keys and re-encrypt the chunks data on the background.
+
+## OUT OF SCOPE
+Integration with external HSM or KMS is out of scope for this delivery. Using a single system-master-key would provide us an easy interface to add it later on since we would need to worry only about our top level master key.
+
+## SOLUTION
+
+### Hierarchy master keys tree
+The following diagram illustrates a simple key hierarchy and its relationship to NooBaa entities data.
+<div id="top" />
+<img src="/docs/design/images/master_keys_diagram.png" />
+In the figure, “Chunk” and “Account” schemas have the data we ultimately want to protect, such as chunk cipher keys, cipher iv, access keys, etc. For these scenarios, symmetric key encryption (where the same key is used both to encrypt and decrypt the data) is the norm, because it’s efficient.
+
+
+The application encrypts the sensitive data using the desired master key, the desired master key itself encrypted by his master key, and so on recursively.
+
+
+When the application wants to decrypt the sensitive data, it fetches the master key and uses it to decrypt the sensitive data.
+
+
+Keys tree is stored encrypted in the DB memory (on disk memory), and stored decrypted at run-time (on process memory) for performance.
```

Comment:
Since the next section describes how we decrypt keys in memory (and we discuss lazy) we don't want to have two different descriptions of it